libkpod/rename.go

```diff
@@ -0,0 +1,108 @@
+package libkpod
+
+import (
+	"encoding/json"
+	"path/filepath"
+
+	"github.com/docker/docker/pkg/ioutils"
+	"github.com/kubernetes-incubator/cri-o/oci"
+	"github.com/opencontainers/runtime-tools/generate"
+)
+
+const (
+	configFile        = "config.json"
+	nameAnnotation    = "io.kubernetes.cri-o.Name"
+	metadatAnnotation = "io.kubernetes.cri-o.Metadata"
+)
+
+// ContainerRename renames the given container
+func (c *ContainerServer) ContainerRename(container, name string) error {
+	ctr, err := c.LookupContainer(container)
+	if err != nil {
+		return err
+	}
+
+	oldName := ctr.Name()
+	_, err = c.ReserveContainerName(ctr.ID(), name)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err != nil {
+			c.ReleaseContainerName(oldName)
+		} else {
+			c.ReleaseContainerName(oldName)
+		}
+	}()
+
+	// Update state.json
+	if err = c.updateStateName(ctr, name); err != nil {
+		return err
+	}
+
+	// Update config.json
+	configRuntimePath := filepath.Join(ctr.BundlePath(), configFile)
+	if err = updateConfigName(configRuntimePath, name); err != nil {
+		return err
+	}
+	configStoragePath := filepath.Join(ctr.Dir(), configFile)
+	if err = updateConfigName(configStoragePath, name); err != nil {
+		return err
+	}
+
+	// Update containers.json
+	return nil
+}
+
+func (c *ContainerServer) updateContainerStoreName(id, name string) error {
+	cstore, err := c.store.ContainerStore()
+	if err != nil {
+		return err
+	}
+
+	return cstore.SetNames(id, []string{name})
+}
+
+func updateConfigName(configPath, name string) error {
+	specgen, err := generate.NewFromFile(configPath)
+	if err != nil {
+		return err
+	}
+	specgen.AddAnnotation(nameAnnotation, name)
+	metadata := specgen.Spec().Annotations[metadatAnnotation]
+	specgen.AddAnnotation(metadata, updateMetadata(metadata, name))
+
+	return specgen.SaveToFile(configPath, generate.ExportOptions{})
+}
+
+func (c *ContainerServer) updateStateName(ctr *oci.Container, name string) error {
+	ctr.State().Annotations[nameAnnotation] = name
+	ctr.State().Annotations[metadatAnnotation] = updateMetadata(ctr.State().Annotations[metadatAnnotation], name)
+	// This is taken directly from c.ContainerStateToDisk(), which can't be used because of the call to UpdateStatus() in the first line
+	jsonSource, err := ioutils.NewAtomicFileWriter(ctr.StatePath(), 0644)
+	if err != nil {
+		return err
+	}
+	defer jsonSource.Close()
+	enc := json.NewEncoder(jsonSource)
+	return enc.Encode(c.runtime.ContainerStatus(ctr))
+}
+
+// Attempts to update a metadata annotation
+func updateMetadata(mdata, name string) string {
+	metadata := make(map[string]interface{})
```

Comment:
Depending on the value of the `annotations.ContainerType` annotation, we know that this is going to be an encoded version of either a `k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime.ContainerMetadata` which we set in `createSandboxContainer()` (called from `CreateContainer()`), or a `k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime.PodSandboxMetadata` which we set in `RunPodSandbox()`.  Decoding using the appropriate type will let us verify that the data decodes into sane values before we attempt to overwrite anything.